node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,796 @@
+	}
+
+	return evt, nil
+}
+
+// parseWasmAttributes parses the attributes in a wasm event. If the contract and action match the desired values (or the desired values are not set)
+// the attributes are loaded into a byte array of the marshaled json. If the event is not of the desired type, it returns nil.
+func parseWasmAttributes(logger *zap.Logger, desiredContract string, desiredAction string, event gjson.Result) ([]byte, error) {
+	attributes := gjson.Get(event.String(), "attributes")
+	if !attributes.Exists() {
+		return nil, fmt.Errorf("message does not contain any attributes")
+	}
+
+	contractAddressSeen := false
+	actionSeen := false
+	attrs := make(map[string]string)
+	for _, attribute := range attributes.Array() {
+		if !attribute.IsObject() {
+			logger.Warn("event attribute is invalid", zap.String("attribute", attribute.String()))
+			continue
+		}
+		keyBase := gjson.Get(attribute.String(), "key")
+		if !keyBase.Exists() {
+			logger.Warn("event attribute does not have key", zap.String("attribute", attribute.String()))
+			continue
+		}
+		valueBase := gjson.Get(attribute.String(), "value")
+		if !valueBase.Exists() {
+			logger.Warn("event attribute does not have value", zap.String("attribute", attribute.String()))
+			continue
+		}
+
+		keyRaw, err := base64.StdEncoding.DecodeString(keyBase.String())
+		if err != nil {
+			logger.Warn("event key attribute is invalid", zap.String("key", keyBase.String()))
+			continue
+		}
+		valueRaw, err := base64.StdEncoding.DecodeString(valueBase.String())
+		if err != nil {
+			logger.Warn("event value attribute is invalid", zap.String("key", keyBase.String()), zap.String("value", valueBase.String()))
+			continue
+		}
+
+		key := string(keyRaw)
+		value := string(valueRaw)
+
+		if key == "_contract_address" {
+			contractAddressSeen = true
+			if desiredContract != "" && value != desiredContract {
+				logger.Debug("ignoring event from an unexpected contract", zap.String("contract", value), zap.String("desiredContract", desiredContract))
+				return nil, nil
+			}
+		} else if key == "action" {
+			actionSeen = true
+			if desiredAction != "" && value != desiredAction {
+				logger.Debug("ignoring event with an unexpected action", zap.String("key", key), zap.String("value", value), zap.String("desiredAction", desiredAction))
+				return nil, nil
+			}
+		} else {
+			if _, ok := attrs[key]; ok {
+				logger.Debug("duplicate key in events", zap.String("key", key), zap.String("value", value))
+				continue
+			}
+
+			logger.Debug("event attribute", zap.String("key", key), zap.String("value", value), zap.String("desiredAction", desiredAction))
+			attrs[string(key)] = value
+		}
+	}
+
+	if !contractAddressSeen && desiredContract != "" {
+		logger.Debug("contract address not specified, which does not match the desired value")
+		return nil, nil
+	}
+
+	if !actionSeen && desiredAction != "" {
+		logger.Debug("action not specified, which does not match the desired value")
+		return nil, nil
+	}
```

Comment:
nit: can we just clarify these comments that:
```
action not specified by wasm event, which does not match the desired value
```

Reply:
That code has been deleted.